narciuch/assignment3/narciuch-3/src/WordSearch.java

```diff
@@ -0,0 +1,145 @@
+   * <li>We cannot visit the same cell more than once in the same word.</li>
+   * </ul>
+   *
+   * @param grid of letters in which to look for words from the dictionary
+   * @param dict the dictionary of words to look for
+   * @return list of words from the dictionary that can be formed in the grid
+   */
+  public List<String> findWords(char[][] grid, Dictionary dict) {
+    List<String> wordsFound = new ArrayList<>();
+    if (grid.length == 0) {
+      return wordsFound;
+    }
+    int nr_rows = grid.length;
+    int nr_columns = grid[0].length;
+    for (int row = 0; row < nr_rows; row++) {
+      for (int column = 0; column < nr_columns; column++) {
+        //Process for each letter in the grid
+
+        //Default boolean value is false - i.e. unvisited letter
+        boolean[][] visited = new boolean[nr_rows][nr_columns];
+
+        List<Character> currentWord = new ArrayList<>();
+
+        findFurtherWords(grid, dict, currentWord, visited, row, column, wordsFound);
+      }
+    }
+    return wordsFound;
+  }
+
+  /**
+   * Helper for findWords.
+   *
+   * <p>Starting at the letter at row 'row' and column 'column' within the grid, this method
+   * recursively checks all possible adjacent* letters in the grid, checking if the already
+   * visited letters on the current path together with the new adjacent letter are a prefix of
+   * a word in the dictionary.
+   *
+   * <p>If they are a prefix, they could form a dictionary word, so we check this and if they do
+   * we add this word to the list of words found.
+   *
+   * <p>If they are a prefix, we must recursively call this method to check if we can form
+   * another prefix or word by adding another adjacent letter to the existing letters.
+   *
+   * <p>*By adjacent letters, we mean adjacent horizontally, vertically or diagonally from the
+   * current letter we are considering.
+   *
+   * <p>When forming words, we assume we cannot visit the same cell more than once in the same word.
+   *
+   * @param grid       the grid of letters we recursively check in
+   * @param dict       the dictionary of words we are looking for
+   * @param letters    the letters we have visited on our current path through the grid
+   * @param visited    an array with the same dimensions as the grid.
+   *                   The boolean entry at row i, column j corresponds to whether we have
+   *                   already visited the letter at row i, column j in the grid, and hence
+   *                   should not revisit it.
+   * @param row        the row of the cell currently being considered as the start point of the possible words
+   * @param column     the column of the cell currently being considered as the start point of the possible words
+   * @param wordsFound list of words we have already been able to form in the grid
+   */
+  private void findFurtherWords(char[][] grid, Dictionary dict, List<Character> letters,
```

Comment:
Code would be simpler if you used `String` instead of `List<Character>` for `letters`. It would remove need to convert list to string and need to remove last element at the end of this method.

Reply:
Agreed. I initially thought it would be more efficient to use a list since we are changing it a lot, but have since realised immutability should be ensured.